tfhe/src/integer/u256.rs

```diff
@@ -344,10 +454,189 @@ impl From<u128> for U256 {
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set & all_odd_bits_set, all_odd_bits_set);
+        assert_eq!(all_even_bits_set & all_even_bits_set, all_even_bits_set);
+        assert_eq!(all_even_bits_set & all_odd_bits_set, U256::ZERO);
+    }
+
+    #[test]
+    fn test_bitor() {
+        let all_even_bits_set = U256([u64_with_even_bits_set(); 4]);
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set | all_odd_bits_set, all_odd_bits_set);
+        assert_eq!(all_even_bits_set | all_even_bits_set, all_even_bits_set);
+        assert_eq!(all_even_bits_set | all_odd_bits_set, U256::MAX);
+    }
+
+    #[test]
+    fn test_bitxor() {
+        let all_even_bits_set = U256([u64_with_even_bits_set(); 4]);
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set ^ all_odd_bits_set, U256::ZERO);
+        assert_eq!(all_even_bits_set ^ all_even_bits_set, U256::ZERO);
+        assert_eq!(all_even_bits_set ^ all_odd_bits_set, U256::MAX);
+    }
+
+    #[test]
+    fn test_is_power_of_two() {
+        assert!(!U256::ZERO.is_power_of_two());
+        assert!(!U256::MAX.is_power_of_two());
+        assert!(!U256::from(8329842348123u64).is_power_of_two());
+
+        assert!(U256::ONE.is_power_of_two());
+        assert!(U256::from(2u64).is_power_of_two());
+        assert!(U256::from(1u64 << 32).is_power_of_two());
+        assert!(U256::from(1u64 << 47).is_power_of_two());
+        assert!(U256::from(1u128 << 84).is_power_of_two());
+    }
+
+    #[test]
+    fn test_ilog2() {
+        assert!(catch_unwind(|| { U256::ZERO.ilog2() }).is_err());
+
+        assert_eq!(U256::MAX.ilog2(), 255);
+        assert_eq!(
+            U256::from(8329842348123u64).ilog2(),
+            8329842348123u64.ilog2()
+        );
+
+        assert_eq!(U256::ONE.ilog2(), 1u32.ilog2());
+        assert_eq!(U256::from(2u64).ilog2(), 2u32.ilog2());
+        assert_eq!(U256::from(1u64 << 32).ilog2(), (1u64 << 32).ilog2());
+        assert_eq!(U256::from(1u64 << 47).ilog2(), (1u64 << 47).ilog2());
+        assert_eq!(U256::from(1u128 << 84).ilog2(), (1u128 << 84).ilog2());
+    }
+
+    #[test]
+    fn test_mul() {
+        let u64_max = U256::from(u64::MAX);
+        let expected = u64::MAX as u128 * u64::MAX as u128;
+        assert_eq!(u64_max * u64_max, U256::from(expected));
+
+        let mut rng = rand::thread_rng();
+        for _ in 0..5 {
+            let a = rng.gen::<u64>();
+            let b = rng.gen::<u64>();
+
+            let res = U256::from(a) * U256::from(b);
+            let expected = a as u128 * b as u128;
+            assert_eq!(res, U256::from(expected));
+        }
```

Comment:
could we have some cases where high bits are set ? maybe with some power of 2 cases ?

Maybe we can re-use some test cases from libs wih some u256 implementations

Reply:
Will need to find examples we can hardcode then

Reply:
Added some hardcoded examples, checked using python

Reply:
awesome thanks !